File: benchmarks/JavaScript/nbody.js
// @ts-check
// The Computer Language Benchmarks Game
// http://shootout.alioth.debian.org/
//
//     contributed by Mark C. Lewis
// modified slightly by Chad Whipkey
//
// Based on nbody.java ported to SOM, and then JavaScript by Stefan Marr.

const { Benchmark } = require('./benchmark');

const PI = 3.141592653589793;
const SOLAR_MASS = 4 * PI * PI;
const DAYS_PER_YER = 365.24;

class Body {
  constructor(x, y, z, vx, vy, vz, mass) {
    this.x = x;
    this.y = y;
    this.z = z;
    this.vx = vx * DAYS_PER_YER;
    this.vy = vy * DAYS_PER_YER;
    this.vz = vz * DAYS_PER_YER;
    this.mass = mass * SOLAR_MASS;
  }

  offsetMomentum(px, py, pz) {
    this.vx = 0.0 - (px / SOLAR_MASS);
    this.vy = 0.0 - (py / SOLAR_MASS);
    this.vz = 0.0 - (pz / SOLAR_MASS);
  }

  static jupiter() {
    return new Body(
      4.84143144246472090e+00,
      -1.16032004402742839e+00,
      -1.03622044471123109e-01,
      1.66007664274403694e-03,
      7.69901118419740425e-03,
      -6.90460016972063023e-05,
      9.54791938424326609e-04
    );
  }

  static saturn() {
    return new Body(
      8.34336671824457987e+00,
      4.12479856412430479e+00,
      -4.03523417114321381e-01,
      -2.76742510726862411e-03,
      4.99852801234917238e-03,
      2.30417297573763929e-05,
      2.85885980666130812e-04
    );
  }

  static uranus() {
    return new Body(
      1.28943695621391310e+01,
      -1.51111514016986312e+01,
      -2.23307578892655734e-01,
      2.96460137564761618e-03,
      2.37847173959480950e-03,
      -2.96589568540237556e-05,
      4.36624404335156298e-05
    );
  }

  static neptune() {
    return new Body(
      1.53796971148509165e+01,
      -2.59193146099879641e+01,
      1.79258772950371181e-01,
      2.68067772490389322e-03,
      1.62824170038242295e-03,
      -9.51592254519715870e-05,
      5.15138902046611451e-05
    );
  }

  static sun() {
    return new Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0);
  }
}
class NBodySystem {
  constructor() {
    this.bodies = this.createBodies();
  }

  createBodies() {
    const bodies = [
      Body.sun(),
      Body.jupiter(),
      Body.saturn(),
      Body.uranus(),
      Body.neptune()
    ];

    let px = 0.0;
    let py = 0.0;
    let pz = 0.0;

    bodies.forEach((b) => {
      px += b.vx * b.mass;
      py += b.vy * b.mass;
      pz += b.vz * b.mass;
    });

    bodies[0].offsetMomentum(px, py, pz);

    return bodies;
  }

  advance(dt) {
    for (let i = 0; i < this.bodies.length; i += 1) {
      const iBody = this.bodies[i];

      for (let j = i + 1; j < this.bodies.length; j += 1) {
        const jBody = this.bodies[j];
        const dx = iBody.x - jBody.x;
        const dy = iBody.y - jBody.y;
        const dz = iBody.z - jBody.z;

        const dSquared = dx * dx + dy * dy + dz * dz;
        const distance = Math.sqrt(dSquared);
        const mag = dt / (dSquared * distance);

        iBody.vx -= dx * jBody.mass * mag;
        iBody.vy -= dy * jBody.mass * mag;
        iBody.vz -= dz * jBody.mass * mag;

        jBody.vx += dx * iBody.mass * mag;
        jBody.vy += dy * iBody.mass * mag;
        jBody.vz += dz * iBody.mass * mag;
      }
    }

    this.bodies.forEach((body) => {
      body.x += dt * body.vx;
      body.y += dt * body.vy;
      body.z += dt * body.vz;
    });
  }

  energy() {
    let e = 0.0;

    for (let i = 0; i < this.bodies.length; i += 1) {
      const iBody = this.bodies[i];
      e += 0.5 * iBody.mass * (iBody.vx * iBody.vx
        + iBody.vy * iBody.vy
        + iBody.vz * iBody.vz);

      for (let j = i + 1; j < this.bodies.length; j += 1) {
        const jBody = this.bodies[j];
        const dx = iBody.x - jBody.x;
        const dy = iBody.y - jBody.y;
        const dz = iBody.z - jBody.z;

        const distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
        e -= (iBody.mass * jBody.mass) / distance;
      }
    }
    return e;
  }
}

class NBody extends Benchmark {
  verifyResult(result, innerIterations) {
    if (innerIterations === 250000) { return result === -0.1690859889909308; }
    if (innerIterations === 1) { return result === -0.16907495402506745; }

    process.stdout.write(`No verification result for ${innerIterations} found\n`);
    process.stdout.write(`Result is: ${result}\n`);
    return false;
  }

  innerBenchmarkLoop(innerIterations) {
    const system = new NBodySystem();
    for (let i = 0; i < innerIterations; i += 1) {
      system.advance(0.01);
    }

    return this.verifyResult(system.energy(), innerIterations);
  }
}

exports.newInstance = () => new NBody();

File: benchmarks/Python/nbody.py
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
#
# Based on nbody.java and adapted based on the SOM version.
from math import sqrt

from benchmark import Benchmark

PI = 3.141592653589793
SOLAR_MASS = 4 * PI * PI
DAYS_PER_YER = 365.24


class _Body:
    def __init__(self, x, y, z, vx, vy, vz, mass):
        self.x = x
        self.y = y
        self.z = z
        self.vx = vx * DAYS_PER_YER
        self.vy = vy * DAYS_PER_YER
        self.vz = vz * DAYS_PER_YER
        self.mass = mass * SOLAR_MASS

    def offset_momentum(self, px, py, pz):
        self.vx = -(px / SOLAR_MASS)
        self.vy = -(py / SOLAR_MASS)
        self.vz = -(pz / SOLAR_MASS)


def jupiter():
    return _Body(
        4.84143144246472090e00,
        -1.16032004402742839e00,
        -1.03622044471123109e-01,
        1.66007664274403694e-03,
        7.69901118419740425e-03,
        -6.90460016972063023e-05,
        9.54791938424326609e-04,
    )


def saturn():
    return _Body(
        8.34336671824457987e00,
        4.12479856412430479e00,
        -4.03523417114321381e-01,
        -2.76742510726862411e-03,
        4.99852801234917238e-03,
        2.30417297573763929e-05,
        2.85885980666130812e-04,
    )


def uranus():
    return _Body(
        1.28943695621391310e01,
        -1.51111514016986312e01,
        -2.23307578892655734e-01,
        2.96460137564761618e-03,
        2.37847173959480950e-03,
        -2.96589568540237556e-05,
        4.36624404335156298e-05,
    )


def neptune():
    return _Body(
        1.53796971148509165e01,
        -2.59193146099879641e01,
        1.79258772950371181e-01,
        2.68067772490389322e-03,
        1.62824170038242295e-03,
        -9.51592254519715870e-05,
        5.15138902046611451e-05,
    )


def sun():
    return _Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0)


class NBodySystem:
    def __init__(self):
        self._bodies = self._create_bodies()

    @staticmethod
    def _create_bodies():
        bodies = [sun(), jupiter(), saturn(), uranus(), neptune()]

        px = 0.0
        py = 0.0
        pz = 0.0

        for b in bodies:
            px += b.vx * b.mass
            py += b.vy * b.mass
            pz += b.vz * b.mass

        bodies[0].offset_momentum(px, py, pz)
        return bodies

    def advance(self, dt):
        for i in range(len(self._bodies)):
            i_body = self._bodies[i]

            for j in range(i + 1, len(self._bodies)):
                j_body = self._bodies[j]

                dx = i_body.x - j_body.x
                dy = i_body.y - j_body.y
                dz = i_body.z - j_body.z

                d_squared = dx * dx + dy * dy + dz * dz
                distance = sqrt(d_squared)
                mag = dt / (d_squared * distance)

                i_body.vx = i_body.vx - (dx * j_body.mass * mag)
                i_body.vy = i_body.vy - (dy * j_body.mass * mag)
                i_body.vz = i_body.vz - (dz * j_body.mass * mag)

                j_body.vx = j_body.vx + (dx * i_body.mass * mag)
                j_body.vy = j_body.vy + (dy * i_body.mass * mag)
                j_body.vz = j_body.vz + (dz * i_body.mass * mag)

        for body in self._bodies:
            body.x = body.x + dt * body.vx
            body.y = body.y + dt * body.vy
            body.z = body.z + dt * body.vz

    def energy(self):
        e = 0.0

        for i in range(len(self._bodies)):
            i_body = self._bodies[i]
            e += (
                0.5
                * i_body.mass
                * (
                    i_body.vx * i_body.vx
                    + i_body.vy * i_body.vy
                    + i_body.vz * i_body.vz
                )
            )

            for j in range(i + 1, len(self._bodies)):
                j_body = self._bodies[j]
                dx = i_body.x - j_body.x
                dy = i_body.y - j_body.y
                dz = i_body.z - j_body.z

                distance = sqrt(dx * dx + dy * dy + dz * dz)
                e -= (i_body.mass * j_body.mass) / distance
        return e


class NBody(Benchmark):
    def inner_benchmark_loop(self, inner_iterations):
        system = NBodySystem()
        for _ in range(inner_iterations):
            system.advance(0.01)

        return self._verify_result(system.energy(), inner_iterations)

    @staticmethod
    def _verify_result(result, inner_iterations):
        if inner_iterations == 250000:
            return result == -0.1690859889909308

        if inner_iterations == 1:
            return result == -0.16907495402506745

        print("No verification result for " + str(inner_iterations) + " found")
        print("Result is: " + str(result))
        return False

    def benchmark(self):
        raise Exception("Should never be reached")

    def verify_result(self, result):
        raise Exception("Should never be reached")
